server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ChunkOutputStream.java

```diff
@@ -0,0 +1,156 @@
+   public void write(int b) {
+      byteSink.emitNext(Tuples.of(new byte[] {(byte)b}, new CompletableFuture<>()), EMIT_FAILURE_HANDLER);
+   }
+
+   @Override
+   public void close() throws IOException {
+       if (!started || byteSink == null) {
+           SinkSubscriber.subscribe(completionSink, Mono.<Void>empty());
+       } else {
+           byteSink.emitComplete(EMIT_FAILURE_HANDLER);
+       }
+   }
+
+   @Override
+   public void write(byte[] bs, int off, int len) {
+       try {
+           asyncWrite(bs, off, len).get();
+       } catch (final InterruptedException ie) {
+           Thread.currentThread().interrupt();
+           throw new RuntimeException(ie);
+       } catch (final ExecutionException ee) {
+           throw new RuntimeException(ee);
+       }
+   }
+
+   @Override
+   public void flush() throws IOException {
+       try {
+           asyncFlush().get();
+       } catch (final InterruptedException ie) {
+           Thread.currentThread().interrupt();
+           throw new RuntimeException(ie);
+       } catch (final ExecutionException ee) {
+           throw new RuntimeException(ee);
+       }
+   }
+
+   @Override
+   public CompletableFuture<Void> asyncFlush() {
+
+       // [AG] Discuss with @crankydillo.  Here is my understanding:
+       //   - asyncFlush is used mainly for SSE.
+       //   - The idea seems to be to send the element immediately without
+       //     waiting for the rest of the elements.  Anyway, that's what SSE is.
+       //   - But, at the same time, it is trying to not overload.  So, kind of
+       //     doing backpressure.  Only if the previous flush is complete,
+       //     then request the next element from the app.
+       //   - The backpressure mechanism is already built into Reactor Netty.  But,
+       //     the question is how to communicate that.
+       //   - Please see https://projectreactor.io/docs/netty/release/reference/index.html#_sse.
+       //     It reads `The flushing strategy is "flush after every element" emitted
+       //     by the provided Publisher`.  So, we do not need to call flush individually.
+       //
+       //   In summary, we are not controlling flushing, instead Reactor Netty does.
+       //   Also, this#asyncWrite will manage the backpressure.  I assume Reactor Netty
+       //   will backpressure if it cannot flush, and that would mean an element won't be
+       //   requested so asyncWrite won't complete until the element is pulled.  Also,
+       //   asyncFlush is called right after an asyncWrite.  So, asyncFlush looks
+       //   useless.
+
+       return CompletableFuture.completedFuture(null);
+   }
+
+   @Override
+   public CompletableFuture<Void> asyncWrite(final byte[] bs, int offset, int length) {
+        final CompletableFuture<Void> cf = new CompletableFuture<>();
+        if (!started) {
+            byteSink = byteSinkSupplier.get();
+            parentResponse.committed();
+            started = true;
+        }
```

Comment:
TBH I'm not entirely sure if it's accessed via multiple threads :) I made the assumption it might be based on the fact that `started` is volatile  and the method name is `asyncWrite`.

Reply:
Where have we ended up on this? Do we know if the output stream is not thread safe?

Reply:
Hi @jamezp .  I think it's fair to say this has not been coded to be thread-safe (at least in terms of 'concurrent' calls).  Our impression is that these methods would not be called concurrently from different threads.  Maybe @asoldano can chime in, but it seems really weird that you'd fire `asyncWrite` from multiple threads without chaining them based on completion of returned `CompletionStage`.  If it's happening 'sequentially' and from different threads, `volatile` should save us.